### PR TITLE
fix: warnUnexpectedRequire binds wrong variable in for-loop

### PR DESCRIPTION
**Pull request by tranvansang**
Fix the warnUnexpectedRequire function where it uses a hoisted variable in a for-loop.

**What kind of change does this PR introduce?**

bugfix

**Did you add tests for your changes?**

No. The bugfix is for a warning. I think tests are unnecessary.

**Does this PR introduce a breaking change?**

No

**What needs to be documented once your changes are merged?**

Nothing.

### CODE REVIEW

File: lib/hmr/JavascriptHotModuleReplacement.runtime.js [hunk 1]
```diff
@@ -107,9 +107,9 @@ module.exports = function () {
 		var outdatedModules = [];
 		var appliedUpdate = {};
 
-		var warnUnexpectedRequire = function warnUnexpectedRequire() {
+		var warnUnexpectedRequire = function warnUnexpectedRequire(moduleId) {
 			console.warn(
-				"[HMR] unexpected require(" + result.moduleId + ") to disposed module"
+				"[HMR] unexpected require(" + moduleId + ") to disposed module"
 			);
 		};
```
Comment (sokra) [inline lib/hmr/JavascriptHotModuleReplacement.runtime.js:114]: As `warnUnexpectedRequire` is a module factory it's parsed a `module` argument, which has an `id` property.

``` js
		var warnUnexpectedRequire = function warnUnexpectedRequire(module) {
			console.warn(
				"[HMR] unexpected require(" + module.id + ") to disposed module"
			);
		};
```

File: lib/hmr/JavascriptHotModuleReplacement.runtime.js [hunk 2]
```diff
@@ -194,7 +194,11 @@ module.exports = function () {
 				}
 				if (doDispose) {
 					addAllToSet(outdatedModules, [result.moduleId]);
-					appliedUpdate[moduleId] = warnUnexpectedRequire;
+					appliedUpdate[moduleId] = warnUnexpectedRequire.bind(
+						null,
+						result.moduleId
+					);
+					appliedUpdate[moduleId].isDisposed = true;
```
Comment (sokra) [inline lib/hmr/JavascriptHotModuleReplacement.runtime.js:201]:
```suggestion
					appliedUpdate[moduleId] = warnUnexpectedRequire;
```

File: lib/hmr/JavascriptHotModuleReplacement.runtime.js [hunk 3]
```diff
@@ -208,7 +212,8 @@ module.exports = function () {
 				$moduleCache$[outdatedModuleId] &&
 				$moduleCache$[outdatedModuleId].hot._selfAccepted &&
 				// removed self-accepted modules should not be required
-				appliedUpdate[outdatedModuleId] !== warnUnexpectedRequire &&
+				(!appliedUpdate[outdatedModuleId] ||
+					!appliedUpdate[outdatedModuleId].isDisposed) &&
```
Comment (sokra) [inline lib/hmr/JavascriptHotModuleReplacement.runtime.js:216]:
```suggestion
				appliedUpdate[outdatedModuleId] !== warnUnexpectedRequire &&
```

**Review by sokra — CHANGES_REQUESTED** [review]
Good find. Code seems correct, but it can be simplified a bit.

A test case would be great.

**Comment by tranvansang** [comment]
@webpack-bot thanks

@sokra I've added a test case that fails without the PR change and succeeds with the PR change. The PR is now available for review.

**Review by sokra — APPROVED** [review]

**Comment by sokra** [comment]
Thanks

**Pull request merged**